cpp/include/raft/matrix/detail/gather.cuh

```diff
@@ -343,6 +349,93 @@ void gather_if(const InputIteratorT in,
   typedef typename std::iterator_traits<MapIteratorT>::value_type MapValueT;
   gatherImpl(in, D, N, map, stencil, map_length, out, pred_op, transform_op, stream);
 }
+
+template <typename MatrixT, typename MapT, typename MapTransformOp, typename IndexT>
+void gatherInplaceImpl(raft::resources const& handle,
+                       raft::device_matrix_view<MatrixT, IndexT, raft::layout_c_contiguous> inout,
+                       raft::device_vector_view<const MapT, IndexT, raft::layout_c_contiguous> map,
+                       MapTransformOp transform_op,
+                       IndexT batch_size)
+{
+  // return type of MapTransformOp, must be convertible to IndexT
+  typedef typename std::result_of<decltype(transform_op)(MapT)>::type MapTransformOpReturnT;
```

Comment:
@cjnolet : Do you know the details of querying the return type of a device function in host code? I remember that there was some issue with it being UB. I don't know if we are running into this issue here. Perhaps there is a way we can avoid it. 

Linking this here:
- https://github.com/NVIDIA/thrust/issues/1688
- https://github.com/NVIDIA/thrust/issues/779

Reply:
I had borrowed this from the old gather implementation [here](https://github.com/rapidsai/raft/blob/6ec78e9ce291737b247e1712c0dc311f8ae4062c/cpp/include/raft/matrix/detail/gather.cuh#L139). In the GoogleTests we are currently not testing if the return type is not convertible to int. But we can check that.

Reply:
Then let's make sure we raise a similar error here if the `transform_op` is not wrapped in `cuda::proclaim_return_type`. That seems to be most safe solution to me. There is no backward-compatibility concern here since gather_inplace is a new function.

Reply:
I think the important piece is to use `cuda::std::result_of`, and libcudacxx will enforce that device lambdas have queryable types at compile time.
```suggestion
  typedef typename cuda::std::result_of<decltype(transform_op)(MapT)>::type MapTransformOpReturnT;
```

Reply:
@bdice can you point me to the header where `cuda::std::result_of` is defined?

Reply:
@ahendriksen I realized that the transform op is actually supposed to be a host functor. Therefore `std::result_of` and `std::is_convertible` does work. I also tried a return type that is not convertible to `int` (`std::string`) and it indeed does throw an exception at that particular line.
Based on suggestions from @bdice I fetched `libcudacxx` from `rapids-cmake`, included the corresponding headers and tried using `cuda::std::result_of` instead of `std::result_of` and I get the following error.
`/nvme/1/tarangj/miniconda3/envs/cuml_test/include/rapids/libcudacxx/cuda/std/detail/libcxx/include/type_traits:4405:16: error: static assertion failed: Attempt to use an extended __device__ lambda in a context that requires querying its return type in host code. Use a named function object, a __host__ __device__ lambda, or cuda::proclaim_return_type instead.`

I therefore believe that we can leave this as is.

Reply:
It seems my previous comment was flawed. The reason `std::result_of` was working was that during testing, we were only using `__host__ __device__` functions. `cuda::std::result_of` adds an extra safety net in that it ensures that the return type of `__device__` only functions is not queried from host code (it throws an error in such a case). The way to avoid this error is by wrapping the function (the `transform_op` here) name inside `cuda::proclaim_return_type<IndexT>` (which will fail if the return type of the `transform_op` is not `IndexT`). This is actually more selective and perhaps what we should be doing. In the current impl, we only check if the return type is convertible to `IndexT`. Now the problem with this is that even double or float is convertible to int, but I'm not sure if the compiler will implicitly do the conversion when we try to use the transformed index (the value returned by the `transform_op`) to obtain the element at a particular position (more specifically, if the returned value has non zero digits after the decimal).

Reply:
Following offline discussions with @cjnolet, the `libcudacxx` dependency has been removed. In addition, the `std::result_of` function calls and checks have been removed. The reasoning is this:
Allowing the whole `transform_op` functor to be its own template gives the user more freedom. Checking if the return type is convertible to `IndexT` is restrictive, which is orthogonal to the freedom given to users. Furthermore, adding `libcu++` as a first class dependency is too big a change for this PR and it can also lead to an inconsistency in the user experience if the `libcu++` related changes are not made elsewhere in `raft`. I have also raised issue https://github.com/rapidsai/raft/issues/1606 to keep track of this.

Reply:
Sounds good to me :+1:. Not adding the libcudacxx dependency in this pull request is a good call.